kernel/sched.c

```diff
@@ -511,53 +534,62 @@ void z_thread_single_abort(struct k_thread *thread)
 			}
 		}
 
-		uint32_t mask = _THREAD_DEAD;
-
-		/* If the abort is happening in interrupt context,
-		 * that means that execution will never return to the
-		 * thread's stack and that the abort is known to be
-		 * complete.  Otherwise the thread still runs a bit
-		 * until it can swap, requiring a delay.
+		/* Wake everybody up who was trying to join with this thread.
+		 * A reschedule is invoked later by k_thread_abort().
 		 */
-		if (IS_ENABLED(CONFIG_SMP) && arch_is_in_isr()) {
-			mask |= _THREAD_ABORTED_IN_ISR;
+		while ((waiter = z_waitq_head(&thread->base.join_waiters)) !=
+		       NULL) {
```

Comment:
Note that wait queues are priority-ordered, so there's a performance cost to using it as the "list backend" for this purpose.  Longer term we might consider unioning a slist or something into the same node memory.

Reply:
Agreed, I think we spoke about this a bit in the original `k_thread_join()` patch, I've filed an enhancement here: https://github.com/zephyrproject-rtos/zephyr/issues/28609